File: Low Level Design (LLD)/Problems/06. Splitwise/Java (with Springboot)/splitwise/src/main/java/com/splitwise/splitwise/service/ExpenseService.java
package com.splitwise.splitwise.service;

import com.splitwise.splitwise.model.Expense;
import com.splitwise.splitwise.model.User;

import java.util.*;

public class ExpenseService {
    private List<Expense> allExpenses = new ArrayList<>();

    public Expense createExpense(Expense expense) {
        expense.calculateSplit();

        for (Map.Entry<User, Double> entry : expense.getUserShare().entrySet()) {
            User user = entry.getKey();
            if (user == null) {
                throw new IllegalArgumentException("Null user in userShare map");
            }
            if (!user.getUserId().equals(expense.getPaidBy().getUserId())) {
                user.updateBalanceSheet(expense.getPaidBy().getUserId(), -entry.getValue());
                expense.getPaidBy().updateBalanceSheet(user.getUserId(), entry.getValue());
            }
        }

        allExpenses.add(expense);
        return expense;
    }

    public List<Expense> getAllExpenses() {
        return allExpenses;
    }
}

File: Low Level Design (LLD)/Problems/06. Splitwise/Cpp/service/src/ExpenseService.cpp
#include "../include/ExpenseService.h"

std::shared_ptr<Expense> ExpenseService::createExpense(const std::shared_ptr<Expense>& expense) {
    expense->calculateSplit();

    const auto& userShare = expense->getUserShare();
    std::shared_ptr<User> paidBy = expense->getPaidBy();

    for (const auto& entry : userShare) {
        const std::shared_ptr<User>& user = entry.first;
        double amountOwed = entry.second;

        if (!user) {
            throw std::invalid_argument("Null user in userShare map");
        }

        if (user->getUserId() != paidBy->getUserId()) {
            user->updateBalanceSheet(paidBy->getUserId(), -amountOwed);
            paidBy->updateBalanceSheet(user->getUserId(), amountOwed);
        }
    }

    allExpenses.push_back(expense);
    return expense;
}

const std::vector<std::shared_ptr<Expense>>& ExpenseService::getAllExpenses() const {
    return allExpenses;
}
